Fixed bounding box handlers not growing and added unit tests

RogPodge opened the pull request:
## Overview
Incorporates suggestion from #7395 to fix bounding box handlers not growing to when the cursor was near them. Also includes refactoring out of date code comments and makes the corners and balls transforms correspond to the actual transforms of the in-editor objects

## Changes
- Fixes: #7395

RogPodge commented:
/azp run

RogPodge commented:
/azp run

RogPodge commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -2259,14 +2258,10 @@ private void HandleProximityScaling()
                 foreach (var pointer in proximityPointers)
                 {
-                    if (IsPointWithinBounds(pointer.Position, maxRadius))
-                    {
```
The one issue I see with removing this check is that we end up calculating the distance of a point to each of the scale handlers, whereas before we could potentially avoid this by only calculating the distance if it was near the bounding box object. However, since we don't actually know the scale of the object that is being manipulated, determining when the pointer is "near" the object becomes kind of tricky. There shouldn't be a large performance impact unless there are a lot of objects with their handlers visible at all the same time.

thalbern replied:
if i remember correctly a performance issue was why troy put this in place - bounding box is heavily used and there can be a lot of handles accumulated in a scene - so removing this check might hit us at a later point

thalbern replied:
can you verify that this doesn't have any performance implications on scenes with many bounding box / bounds control added ?

RogPodge replied:
With 128~ bounding boxes in the scene I go from 29 to around 24 fps

With 65~ ish bounding boxes I go from 35-37 fps to around 33

With 32~ ish bounding boxes its 102 with the check and 99 without

thalbern commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -1502,7 +1502,7 @@ private void AddCorners()
 
                 AddComponentsToAffordance(corner, new Bounds(cornerbounds.center * invScale, cornerbounds.size * invScale), RotationHandlePrefabCollider.Box, CursorContextInfo.CursorAction.Scale, scaleHandleColliderPadding);
-                corners.Add(corner.transform);
+                corners.Add(cornerVisual.transform);
```
there's two things about this:
1. we're trying to get away from bounding box and use bounds control instead - fixes made in boundingbox should also be made in boundscontrol (in non critical cases I would even recommend doing the changes only in bounds control to avoid having to fix them in two spots)
2. I'm very certain that this will break a lot of the boundscontrol behavior on reloading visuals on config changes - i don't have all these changes published yet but this definitely will have consequences on all the rig reloading :-/

RogPodge replied:
I'll look into ways to avoid changing the behavior of the corners variable then

thalbern replied:
resolving this as we had a offline chat about this - boundscontrol behavior will be checked separately

thalbern commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -2410,14 +2392,14 @@ private HandleType GetHandleType(Transform handle)
             for (int i = 0; i < balls.Count; ++i)
             {
-                if (handle == balls[i])
+                if (handle.position == balls[i].position && handle.rotation == balls[i].rotation)
```
so had a second look now and I'm not sure I understand why you had to change to the visuals here instead of sticking to the handle object itself? The handle object is a parent of the visual, so it should be fine to use it's transforms and it's also making these checks less error prone as you're comparing the object and not just some arbitrary position / rotation values

RogPodge replied:
The main reason for this change was for testing purposes, as the only public way to access handlers was through the ScaleCorners and RotateMidpoints objects, which refer to the corners and balls objects. I can rework ScaleCorners and RotateMidpoints to directly refer to the handler objects themselves though

RogPodge commented:
I reverted some of my API changes but also fixed the bounds check. Previously we assumed that the center of a bounding box was at transform.position, but we don't account for when the box collider is offset from the center of the transform. I've made the fix for both the old bounding box and the new bounds control

thalbern commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -2252,9 +2274,8 @@ private void HandleProximityScaling()
-                maxRadius *= maxRadius;
-                maxRadius += handleCloseProximity + handleMediumProximity;
+                float maxRadius = currentBoundsExtents.sqrMagnitude;
+                maxRadius += 3 * handleMediumProximity * handleMediumProximity;
```
can you explain why this is now squared length and triple medium proximity  added?

thalbern replied:
without a comment this seems pretty arbitrary just to make it work?

RogPodge replied:
I was copying what we did in bounds control 

https://github.com/microsoft/MixedRealityToolkit-Unity/blob/18fe32fd8d627216754885c7fecf4cf2434aa07d/Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/Visuals/ProximityEffect/ProximityEffect.cs#L153

thalbern replied:
ha! i'm glad i left that comment there - could you please copy over the comment from bounds control two lines above so it's more clear what this does :D

thalbern commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -1148,7 +1148,18 @@ public IReadOnlyList<Transform> ScaleCorners
+        /// <summary>
+        /// Returns list of transforms pointing to the scale handle visuals of the bounding box.
+        /// </summary>
+        public IReadOnlyList<Transform> ScaleCornerVisuals
```
i might be missing some important detail here: I can't spot you're using this anywhere in your fix. why are you adding those visuals and also exposing them to public API? If this is because of the tests you shouldn't add things to the public API of a sdk component just for test reasons - you can access these gameobjects as well by their name for testing purposes :)

RogPodge replied:
I didn't really want to access by name (searching for corners_i and then searching for the visuals game object under that object's children seems less ideal and more prone to breakage)

thalbern commented on Assets/MRTK/SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:
```diff
@@ -2373,7 +2393,7 @@ private void ScaleHandle(Handle handle, bool lerp = false)
         private bool IsPointWithinBounds(Vector3 point, float radiusSqr)
         {
-            return (transform.position - point).sqrMagnitude < radiusSqr;
+            return (Vector3.Scale(TargetBounds.center, TargetBounds.gameObject.transform.lossyScale) + transform.position - point).sqrMagnitude < radiusSqr;
```
is targetbounds center relative to the transform of the component? or is this returning world space coordinates?

RogPodge replied:
I was copying what we did in bounds control 

https://github.com/microsoft/MixedRealityToolkit-Unity/blob/18fe32fd8d627216754885c7fecf4cf2434aa07d/Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/Visuals/ProximityEffect/ProximityEffect.cs#L153

thalbern replied:
I wasn't sure if the center of those bounds is in world or local space - makes sense in local space, closing this issue

thalbern commented on Assets/MRTK/Tests/PlayModeTests/BoundingBoxTests.cs:
```diff
@@ -109,6 +109,67 @@ public IEnumerator BBoxOverride()
+            var edgeHandlerPosition = bbox.RotateMidpoints[0].transform.position;
+
+            // Wait for the scaling/unscaling animation to finish
+            yield return new WaitForSeconds(0.2f);
```
if you modify the CloseGrowRate / FarGrowRate / MediumGrowRate and set them to 1.0f in test scenario you don't have to wait for the animation to finish and the scaling changes should be applied immediately

RogPodge replied:
Those variables aren't publicly modifiable so will be leaving this.

thalbern commented on Assets/MRTK/Tests/PlayModeTests/BoundingBoxTests.cs:
```diff
@@ -109,6 +109,67 @@ public IEnumerator BBoxOverride()
+            var inputSimulationService = PlayModeTestUtilities.GetInputSimulationService();
+
+            // Definining the edge and corner handlers that will be used
+            var originalCornerHandlerScale = bbox.ScaleCornerVisuals[0].transform.localScale;
```
getting a scale handle can be achieved by calling for example: transform.Find("rigRoot/corner_0") and it's visual by calling transform.Find("rigRoot/corner_0/visualsScale/visuals") on the boundingbox gameobject
if you want to have the front right corner that would be corner_3 . hope that helps :)

RogPodge commented:
/azp run

thalbern commented:
one last thing. can you retarget this to stabilization branch? it's a ticket for 2.4 version :)

thalbern commented:
/azp run

RogPodge commented:
/azp run
